File: src/main/scala/org/araqnid/typevariance/scala/InvariantEventDeserializer.scala
package org.araqnid.typevariance.scala

import org.araqnid.typevariance.{BasicEvent, Event}

trait InvariantEventDeserializer[E] {
  def deserialize(bytes: Array[Byte]): E
}

object InvariantDeserializers {
  val invariantEventDeserializer: InvariantEventDeserializer[_ <: Event] = new InvariantEventDeserializer[Event] {
    override def deserialize(bytes: Array[Byte]): Event = ???
  }

  val invariantBasicEventDeserializer: InvariantEventDeserializer[_ <: Event] = new InvariantEventDeserializer[BasicEvent] {
    override def deserialize(bytes: Array[Byte]): BasicEvent = ???
  }
}

File: src/main/kotlin/org/araqnid/typevariance/kotlin/InvariantEventDeserializer.kt
package org.araqnid.typevariance.kotlin

import org.araqnid.typevariance.BasicEvent
import org.araqnid.typevariance.Event

interface InvariantEventDeserializer<E> {
    fun deserialize(bytes: ByteArray): E
}

val invariantEventDeserializer: InvariantEventDeserializer<out Event> = object : InvariantEventDeserializer<Event> {
    override fun deserialize(bytes: ByteArray): Event = TODO()
}

val invariantBasicEventDeserializer: InvariantEventDeserializer<out Event> = object : InvariantEventDeserializer<BasicEvent> {
    override fun deserialize(bytes: ByteArray): BasicEvent = TODO()
}
